source/server/connection_handler_impl.cc

```diff
@@ -148,6 +150,11 @@ void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
       // prevent further redirection.
       new_listener->onAccept(std::move(socket_), false);
     } else {
+      // Set default transport protocol if none of the listener filters did it.
+      if (socket_->detectedTransportProtocol().empty()) {
+        socket_->setDetectedTransportProtocol(
```

Comment:
Should we at least add an ASSERT inside the implementation of `setDetectedTransportProtocol` that it's not set multiple times?

Reply:
I've been going back and forth on this, really.

I expect that sooner or later we'll have multiple transport sockets implementing the same transport protocol (e.g. we could have builtin BoringSSL implementation and another implementation with HSM integration, both using `"tls"` transport protocol). In theory, single TLS Inspector is all that's needed, but I expect we're going to see "sniffer + transport socket" combos, once the ecosystem of custom extensions starts growing up.

Reply:
That makes sense, I just wonder if we should add the ASSERT for now for sanity. We can always remove it later. I don't have a strong opinion either way.

Reply:
Let's skip it -`ASSERT()` on 2nd `setDetectedTransportProtocol()` might result in hard to debug crashes, is going to be removed at some point anyway, and it might prevent people from deploying custom and/or private sniffers.